server/src/main/java/org/elasticsearch/common/file/AbstractFileWatchingService.java

```diff
@@ -293,12 +293,20 @@ final WatchKey enableDirectoryWatcher(WatchKey previousKey, Path settingsDir) th
     void processSettingsAndNotifyListeners() throws InterruptedException {
         try {
             processFileChanges();
-            for (var listener : eventListeners) {
-                listener.watchedFileChanged();
-            }
         } catch (IOException | ExecutionException e) {
-            logger.error(() -> "Error processing watched file: " + watchedFile(), e);
+            onProcessFileChangesException(e);
+            return;
         }
+        for (var listener : eventListeners) {
+            listener.watchedFileChanged();
+        }
```

Comment:
This loop can't throw `IOException` or `ExecutionException` anyhow. Move them out of the `try` block for clarity.